### ISSUE TITLE
Cast -9058 float to unsigned mysql compatibility

### ISSUE THREAD

**Issue by DQinYuan** [pingcap/tidb#12436]
## Bug Report

1. What did you do?

```sql
select cast(cast(-9058 as float) as unsigned);
show warnings;
```

2. What did you expect to see?

In mysql 8.0, no warnings or errors are produced:

```sql
mysql [localhost:8022] {msandbox} (test) > select cast(cast(-9058 as float) as unsigned);
+----------------------------------------+
| cast(cast(-9058 as float) as unsigned) |
+----------------------------------------+
|                   18446744073709542558 |
+----------------------------------------+
1 row in set (0.00 sec)

mysql [localhost:8022] {msandbox} (test) > show warnings;
Empty set (0.00 sec)
```

3. What did you see instead?

In TiDB, results are correct but there is a warning:

```sql
mysql> select cast(cast(-9058 as float) as unsigned);
+----------------------------------------+
| cast(cast(-9058 as float) as unsigned) |
+----------------------------------------+
|                   18446744073709542558 |
+----------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------+
| Level   | Code | Message                         |
+---------+------+---------------------------------+
| Warning | 1690 | constant -9058 overflows bigint |
+---------+------+---------------------------------+
1 row in set (0.00 sec)

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```sql
mysql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-1742-gf51bdfc51-dirty
Edition: Community
Git Commit Hash: f51bdfc514ed48a288922526dffe9e4b234cc788
Git Branch: master
UTC Build Time: 2020-12-07 12:20:21
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)

```

## SIG slack channel

 [#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score

300

## Mentor

- @lzmhhh123

**Comment by nullnotnil** [comment]
Confirming that I can reproduce this in master - except with a warning and not an error:

```sql
drop table if exists ftest;
create table ftest (f float);
insert into ftest values(-9058);
select cast(f as unsigned) from ftest;
show warnings;


..

mysql> select cast(f as unsigned) from ftest;
+----------------------+
| cast(f as unsigned)  |
+----------------------+
| 18446744073709542558 |
+----------------------+
1 row in set, 1 warning (0.00 sec)

mysql> 
mysql> show warnings;
+---------+------+---------------------------------+
| Level   | Code | Message                         |
+---------+------+---------------------------------+
| Warning | 1690 | constant -9058 overflows bigint |
+---------+------+---------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-771-gca41972fb
Edition: Community
Git Commit Hash: ca41972fbac068c8a5de107d9075f09ac68842ac
Git Branch: master
UTC Build Time: 2020-07-14 02:41:21
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)

```

**Comment by wjhuang2016** [comment]
@morgo I think we can accept the warning, what do you think?

**Comment by morgo** [comment]
I think it's worth fixing, but not high priority. Here is a reduced testcase:
```sql
select cast(cast(-9058 as float) as unsigned);
show warnings;
..

mysql> select cast(cast(-9058 as float) as unsigned);
+----------------------------------------+
| cast(cast(-9058 as float) as unsigned) |
+----------------------------------------+
|                   18446744073709542558 |
+----------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------+
| Level   | Code | Message                         |
+---------+------+---------------------------------+
| Warning | 1690 | constant -9058 overflows bigint |
+---------+------+---------------------------------+
1 row in set (0.00 sec)
```

I am going to update the original description and title.

**Comment by wjhuang2016** [comment]
The previous bug had been fixed in both v4.0.8 and master.
